packages/contractkit/src/identity/offchain/schema-utils.ts

```diff
@@ -21,62 +29,399 @@ export class OffchainError extends RootError<SchemaErrorTypes.OffchainError> {
   }
 
   read = makeAsyncThrowable(this.readAsResult.bind(this))
 
-  async write(data: T) {
-    return writeWithSchema(this.wrapper, this.type, this.dataPath, data)
+  private async readEncrypted(account: string): Promise<Result<DataType, SchemaErrors>> {
+    const encryptedResult = await readEncrypted(this.wrapper, this.dataPath, account)
+
+    if (encryptedResult.ok) {
+      return this.deserialize(encryptedResult.result)
+    }
+
+    return encryptedResult
   }
 }
 
-export const readWithSchemaAsResult = async <T>(
-  wrapper: OffchainDataWrapper,
-  type: t.Type<T>,
-  account: Address,
-  dataPath: string
-): Promise<Result<T, SchemaErrors>> => {
-  const resp = await wrapper.readDataFromAsResult(account, dataPath)
+export class BinarySchema {
+  constructor(readonly wrapper: OffchainDataWrapper, readonly dataPath: string) {}
+
+  async write(data: Buffer) {
+    const signature = await signBuffer(this.wrapper, this.dataPath, data)
+    await this.wrapper.writeDataTo(data, signature, this.dataPath)
+  }
+
+  async writeEncrypted(data: Buffer, toAddress: string) {
+    return writeEncrypted(this.wrapper, this.dataPath, data, toAddress)
+  }
 
-  if (!resp.ok) {
-    return Err(new OffchainError(resp.error))
+  async writeWithSymmetric(data: Buffer, toAddresses: string[], symmetricKey?: Buffer) {
+    return writeEncryptedWithSymmetric(this.wrapper, this.dataPath, data, toAddresses, symmetricKey)
   }
 
-  try {
-    const asJson = JSON.parse(resp.result)
-    const parseResult = type.decode(asJson)
-    if (isRight(parseResult)) {
-      return Ok(parseResult.right)
+  async readAsResult(account: string): Promise<Result<Buffer, SchemaErrors>> {
+    const rawData = await this.wrapper.readDataFromAsResult(
+      account,
+      (buf) => buildBinaryEIP712TypedData(this.wrapper, this.dataPath, buf),
+      this.dataPath
+    )
+    if (!rawData.ok) {
+      return this.readEncrypted(account)
     }
-    return Err(new RootError(SchemaErrorTypes.InvalidDataError))
-  } catch (error) {
-    return Err(new RootError(SchemaErrorTypes.InvalidDataError))
+
+    const encryptedResult = await this.readEncrypted(account)
+
+    if (encryptedResult.ok) {
+      return encryptedResult
+    }
+    return Ok(rawData.result)
+  }
+
+  read = makeAsyncThrowable(this.readAsResult.bind(this))
+
+  private async readEncrypted(account: string) {
+    return readEncrypted(this.wrapper, this.dataPath, account)
   }
 }
 
-export const readWithSchema = makeAsyncThrowable(readWithSchemaAsResult)
+// label = PRF(ECDH(A, B), A || B || data path)
+// ciphertext path = "/cosmetic path/" || base64(label)
+export function getDataPath(
+  path: string,
+  sharedSecret: Buffer,
+  senderPublicKey: string,
+  receiverPublicKey: string
+) {
+  const senderPublicKeyBuffer = Buffer.from(trimLeading0x(senderPublicKey), 'hex')
+  const receiverPublicKeyBuffer = Buffer.from(trimLeading0x(receiverPublicKey), 'hex')
+
+  return createHmac('sha3-256', sharedSecret)
```

Comment:
After some additional feedback from Psi we decided to switch to BLAKE2s
```suggestion
  return createHmac('blake2s256', sharedSecret)
```